### PR TITLE
Implement volumetric fog and volumetric lighting, also known as light shafts or god rays.

### PR DESCRIPTION
**Pull request by pcwalton**
This commit implements a more physically-accurate, but slower, form of fog than the `bevy_pbr::fog` module does. Notably, this *volumetric fog* allows for light beams from directional lights to shine through, creating what is known as *light shafts* or *god rays*.

To add volumetric fog to a scene, add `VolumetricFogSettings` to the camera, and add `VolumetricLight` to directional lights that you wish to be volumetric. `VolumetricFogSettings` has numerous settings that allow you to define the accuracy of the simulation, as well as the look of the fog. Currently, only interaction with directional lights that have shadow maps is supported. Note that the overhead of the effect scales directly with the number of directional lights in use, so apply `VolumetricLight` sparingly for the best results.

The overall algorithm, which is implemented as a postprocessing effect, is a combination of the techniques described in [Scratchapixel] and [this blog post]. It uses raymarching in screen space, transformed into shadow map space for sampling and combined with physically-based modeling of absorption and scattering. Bevy employs the widely-used [Henyey-Greenstein phase function] to model asymmetry; this essentially allows light shafts to fade into and out of existence as the user views them.

Volumetric rendering is a huge subject, and I deliberately kept the scope of this commit small. Possible follow-ups include:

1. Raymarching at a lower resolution.

2. A post-processing blur (especially useful when combined with (1)).

3. Supporting point lights and spot lights.

4. Supporting lights with no shadow maps.

5. Supporting irradiance volumes and reflection probes.

6. Voxel components that reuse the volumetric fog code to create voxel shapes.

7. *Horizon: Zero Dawn*-style clouds.

These are all useful, but out of scope of this patch for now, to keep things tidy and easy to review.

A new example, `volumetric_fog`, has been added to demonstrate the effect.

### Added

* A new component, `VolumetricFog`, is available, to allow for a more physically-accurate, but more resource-intensive, form of fog.

* A new component, `VolumetricLight`, can be placed on directional lights to make them interact with `VolumetricFog`. Notably, this allows such lights to emit light shafts/god rays.

Screenshot 2024-04-21 162808

Screenshot 2024-04-21 132005

[Scratchapixel]: https://www.scratchapixel.com/lessons/3d-basic-rendering/volume-rendering-for-developers/intro-volume-rendering.html

[this blog post]: https://www.alexandre-pestana.com/volumetric-lights/

[Henyey-Greenstein phase function]: https://www.pbr-book.org/4ed/Volume_Scattering/Phase_Functions#TheHenyeyndashGreensteinPhaseFunction

### CODE REVIEW

File: crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl
Comment (mweatherley) [inline crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl:158]: Just double-checking my understanding here, but my impression is that one of the primary consequences of using a directional light (putting aside technical matters like sampling the shadow map) is that this phase parameter is constant throughout the raymarching process?

File: crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl
Comment (mweatherley) [inline crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl:126]: I think I get the basic idea of this hack, but I'm still struggling a little bit to understand exactly what this attenuation represents; based on the "not needing to shoot secondary rays toward the directional light" bit, I'm imagining that this is supposed to represent, for each point in the raymarch, the light attenuation of traveling from the directional light to the sample point? If so, I think that could be a little clearer.

(That would also clear up my confusion over the line following this, where the density, step size, and scattering coefficient appear once again — because they are for the attenuation on the journey to the camera from the sample point, I think.)

**Review by mweatherley — APPROVED** [review]
I'm not really qualified to comment on the changes to view layouts or the rendering pass setup, but the shader algorithm looks good to me, and I have run the example locally with success. Just a few minor things about potentially clearing a few things up.

File: crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl:153]: The depth bias is to avoid shadow acne. Google pulls this up: https://digitalrune.github.io/DigitalRune-Documentation/html/3f4d959e-9c98-4a97-8d85-7a73c26145d7.htm

File: crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl:60]: These are the standard names in the literature and will be familiar to graphics programmers. https://en.wikipedia.org/wiki/Blinn%E2%80%93Phong_reflection_model

1920px-Blinn_Vectors svg

File: crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl:195]: I added a comment.

File: crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl]: Done.

File: crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl:158]: Yep. We can lift its calculation out of the loop, which is nice because Henyey-Greenstein is kind of expensive to evaluate.

File: crates/bevy_pbr/src/volumetric_fog/mod.rs
Comment (pcwalton) [reply crates/bevy_pbr/src/volumetric_fog/mod.rs:148]: Scattering is both towards and away from the viewer. Note that `scattering` is a factor in both out-scattering (`sample_attenuation`) and in-scattering (`light_factors_per_step` -> `light_color_per_step` -> `accumulated_color`). The difference between out-scattering and in-scattering is that in-scattering is affected by phase (i.e. Henyey-Greenstein). If not for phase, the amount of in-scattering and out-scattering would be the same.

In other words, `scattering` provides a *baseline* level of scattering for both in-scattering and out-scattering, while `phase` determines the proportion that goes toward the viewer as opposed to away.

File: crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl:126]: Yes. I reworded the comment.

**Comment by pcwalton** [comment]
I think the CI failure is spurious.

**Comment by BD103** [comment]
> I think the CI failure is spurious.

Yup, feel free to rebase / merge from main to get the check started again. :)

File: crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl
Comment (mweatherley) [reply crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl:60]: My comment is regarding the fact that `LdotV` is the negative of the cosine parameter in the linked documentation, which led me to believe that the V being supplied is actually the negative of the one in that image, but maybe I'm mistaken.

**Comment by pcwalton** [comment]
Closing as I am burned out and have no motivation to get this landed anymore.

**Review by Aceeri — APPROVED** [review]
Minor nitpick and a little documentation fix, but overall looks good to me!

File: crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl:60]: You're right, that's actually -(L dot V). I renamed it to `neg_LdotV`.

File: crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl
Comment (IceSentry) [inline crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl:89]: This won't work on webgl but that's not really a big issue since this feature is optional anyway. We might want to warn about that inside the plugin initialization. Not a blocker for merging though. Also, the fact that we can't sample depth in webgl is most likely a wgpu issue and not a webgl issue.

**Review by IceSentry — APPROVED** [review]
The plugin doc should probably have a warning that this won't work in webgl and potentially a logged warning message on plugin building too.

Other than that LGTM

**Comment by IceSentry** [comment]
Note to maintainers. I extracted the mesh view layouts change to it's own PR since those changes make sense outside of this PR. #13266 

That PR should be merged first, but it's not a blocker if we don't do that.

**Comment by IceSentry** [comment]
@pcwalton since https://github.com/bevyengine/bevy/pull/13266 was merged this needs a merge. There won't be conflict, but you'll be able to remove the `.view_layouts.get_view_layout` that you added in a few places.

**Comment by alice-i-cecile** [comment]
Once merge conflicts are resolved I'll merge this in :)

**Comment by IceSentry** [comment]
I'm not sure I did it correctly, but I merged main and fixed the confict in this PR to your PR https://github.com/pcwalton/bevy/pull/5

**Comment by pcwalton** [comment]
This should be ready now.

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1318 if you'd like to help out.

**Pull request merged**
